api/v1alpha1/resource_error.go

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2022 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1alpha1
+
+type ResourceError struct {
+	// Optional user facing message if the error is relevant to an end user
+	UserMessage string `json:"userMessage,omitempty"`
+
+	// Internal debug message for the error
+	DebugMessage string `json:"debugMessage"`
+
+	// Indication if the error is likely recoverable or not
+	// +kubebuilder:validation:Default=true
+	Recoverable bool `json:"recoverable"`
+}
+
+func NewResourceError(message string, err error) *ResourceError {
+	resourceError := &ResourceError{
+		Recoverable: true,
+	}
+
+	if err != nil {
+		// If the error provided is already a ResourceError, use it and concatenate
+		// the debug messages
+		_, ok := err.(*ResourceError)
+		if ok {
+			resourceError = err.(*ResourceError)
+		}
+
+		if message == "" {
+			message = err.Error()
+		} else {
+			message = message + ": " + err.Error()
+		}
+	}
+
+	resourceError.DebugMessage = message
+
+	return resourceError
+}
+
+func (e *ResourceError) GetDebugMessage() string {
```

Comment:
Why do you need Getters & Setters if the members are public? Should they be private?

Reply:
I just did that out of habit. I don't think you can have private fields in a struct that's part of a k8s resource. I can remove them.